File: stream/src/event.ts
/**
 * @license Apache-2.0
 */

/**
 * Currency pair.
 */
export interface Pair {
  quote: string;
  base: string;
}

/**
 * Market stream event.
 */
export interface Events {
  seq: number;
  market: string;
  pair: Pair;
  events: Event[];
}

/**
 * Market event.
 */
export interface Event {
  Trade?: Trade;
  Order?: Order;
}

/**
 * Market order.
 */
export interface Order {
  kind: 'Ask' | 'Bid';
  rate: number;
  volume: number;
  total?: number;
}

/**
 * Market trade.
 */
export interface Trade {
  id: number;
  order: Order;
  timestamp: number;
}

File: stream/src/event.rs
//! Market stream event.

/// Market stream events.
#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
pub struct Events {
    /// Sequence number.
    pub seq: Option<i64>,

    /// Market identifier.
    pub market: ::market::MarketId,

    /// Currency pair.
    pub pair: Option<::currency::Pair>,

    /// Vector of events.
    pub events: Vec<::market::Event>,

    /// Timestamp of market event.
    pub timestamp: Option<i64>,
}

/// Market stream event sender.
pub type EventSender = ::futures::sync::mpsc::UnboundedSender<Events>;

/// Market stream event receiver.
pub type EventReceiver = ::futures::sync::mpsc::UnboundedReceiver<Events>;
